src/MySqlConnector/Serialization/MySqlSession.cs

```diff
@@ -35,16 +36,97 @@ public MySqlSession(ConnectionPool pool, int poolGeneration)
 		public ConnectionPool Pool { get; }
 		public int PoolGeneration { get; }
 		public string DatabaseOverride { get; set; }
+		public IPAddress IPAddress => (m_tcpClient?.Client.RemoteEndPoint as IPEndPoint)?.Address;
 
 		public void ReturnToPool() => Pool?.Return(this);
 
-		public bool IsConnected => m_state == State.Connected;
+		public bool IsConnected
+		{
+			get
+			{
+				lock (m_lock)
+					return m_state == State.Connected;
+			}
+		}
+
+		public bool TryStartCancel(MySqlCommand command)
+		{
+			lock (m_lock)
+			{
+				if (m_activeCommand != command)
+					return false;
+				VerifyState(State.Querying, State.CancelingQuery);
+				if (m_state != State.Querying)
+					return false;
+				m_state = State.CancelingQuery;
+			}
+
+			return true;
+		}
+
+		public void DoCancel(MySqlCommand commandToCancel, MySqlCommand killCommand)
+		{
+			lock (m_lock)
+			{
+				if (m_activeCommand != commandToCancel)
+					return;
+
+				try
+				{
+					killCommand.ExecuteNonQuery();
```

Comment:
> wanted to avoid the following problem...

Ah, that is a good point.  

> We could dial down the `CommandTimeout`

Maybe should also dial down the `ConnectionTimeout` for the cancellation query.  I believe `CommandTimeout` will simply be calling this `.Cancel()` method, right?